Add a document describing how to triage CI test failures

BruceForstall opened the pull request.

BruceForstall committed:
Add a document describing how to triage CI test failures

BruceForstall commented:
@jkotas @hoyosjs @agocke @danmoseley @dotnet/jit-contrib I wrote a document on triaging CI test failures, especially with the goal of distinguishing between JIT stress and non-stress failures. Is this useful? If so, how could it be better?

agocke commented:
Looks good, but here:

> For intermittent failures, this might require running the test in a loop, applying artificial load to the machine, running on the exact architecture/OS/OS version/Docker container, etc.

this is a pretty involved investigation. I can't see many people outside the JIT going through that much effort for a single failure. It seems like the value of the stress leg is in finding issues that wouldn't occur in normal configurations, so if the result of a failure is trying to run a stress test locally, that seems like a failure of the infrastructure.

hoyosjs commented:
Overall, I think it's a good idea to have a document that has things that can help properly route an issue. It helps people that come to the team, and it helps people who have an issue that hits CI. I feel like it was long overdue and 

I don't know if this is really answers stress/non-stress issue, other than "try to repro, and if removing the variable no longer repros - then it's likely stress". A good query gives you large numbers of runs. We should leverage that in those cases (basically - is a relevant portion of these seen outside of stress legs? Otherwise it's very likely something that's stress related. or stress incompatible).

danmoseley commented on docs/workflow/ci/triaging-failures.md:
```diff
@@ -0,0 +1,137 @@
+Then use the `runfo get-helix-payload` command.
+
+Note that if a test fails and produces a core (crash) dump, the Azure DevOps "Artifacts" page will include a link to
+the crash dump as well as a `how-to-debug-dump.md` file that describes using `runfo` to download the assets and other
```
it would be good to make it work for clr tests too, of course. there's nothing libraries specific about it.

hoyosjs replied:
I've been meaning to hook it up along with a couple of other nice tricks.

danmoseley commented on docs/workflow/ci/triaging-failures.md:
```diff
@@ -0,0 +1,137 @@
+
+### Intermittent test failures
+
+If it is difficult to reproduce an issue, a useful tool is to use [Kusto](https://dataexplorer.azure.com/clusters/engsrvprod/databases/engineeringdata)
```
this DB is currently only accessible to Microsoft employees, unfortunately.  (not sure about runfo?)

danmoseley commented:
I think it's a useful start -- thanks @BruceForstall for kicking it off. It should probably be linked from some other doc(s).

Is it interesting to describe the various jit stress modes? For example, what are the slowest ones?

Where does GC stress fit here, is it relevant?

danmoseley commented on docs/workflow/ci/triaging-failures.md:
```diff
@@ -0,0 +1,137 @@
+
+### Intermittent test failures
+
+If it is difficult to reproduce an issue, a useful tool is to use [Kusto](https://dataexplorer.azure.com/clusters/engsrvprod/databases/engineeringdata)
```
When a test fails in Jitstress, would it make sense to ask the opener to run and paste in the results of a canned Kusto query to see whether it's only failing in Jitstress? Then they could decide whether to send it to codegen team first or not. (Of course, it may well be a timing issue, but presumably this heuristic would be more accurate than always sending to the area owners first)

danmoseley replied:
It might be interesting to do a hand sampling of the last 20-50 bugs marked jit stress to see what proportion were actually test issues (not codegen team issues), and of those how many were only failing in jitstress. That could help us tune our process so they go to the right team first.

hoyosjs replied:
Yes, this is what I tend to use to first rule it out. No need to tag jit stress if it reproduces obviously outside the range (I say a significant portion because one or two failures outside usually hints at other issues like a run with something more systemic and less likely to be the test itself).

BruceForstall commented:
> this is a pretty involved investigation. I can't see many people outside the JIT going through that much effort for a single failure. It seems like the value of the stress leg is in finding issues that wouldn't occur in normal configurations, so if the result of a failure is trying to run a stress test locally, that seems like a failure of the infrastructure.

My point here is that a test can, and very often does, fail to reproduce without these specific attempts, *independent* of stress mode. If a bug can't be reproduced with the stress mode where it originally failed (with the precise environment), then we can't investigate the failure. However, that is true no matter the stress mode.

BruceForstall committed:
Updated for feedback

1. Added a section about "finding the most general failure configuration"
2. Moved Kusto section up, to suggest using that to determine breadth of
failures
3. Added more about GCStress, crossgen2, ilasm round-trip testing modes

BruceForstall commented:
I updated this doc. @hoyosjs @agocke @danmoseley Another round of review?

danmoseley commented on docs/workflow/ci/triaging-failures.md:
```diff
@@ -0,0 +1,202 @@
+First, identify the platform the test was run on:
+1. The processor architecture (x86, x64, arm32, arm64)
+2. The operating system (Windows, Linux, macOS). In some cases, you might need to reproduce a failure under a specific operating system
+version, such as a "musl" version of Linux (e.g., Alpine). You might need to reproduce using the same environment used by the CI system.
```
```suggestion
version: for example, a failure might be specific to Alpine, which uses a different C runtime than most distros. You might need to reproduce using the same environment used by the CI system.
```

danmoseley commented on docs/workflow/ci/triaging-failures.md:
````diff
@@ -0,0 +1,202 @@
+To install `runfo` as a .NET CLI global tool:
+
+```sh
+dotnet tool install --global runfo
````
Seems to me that if I do "dotnet tool install" and it's already installed, it ought to implicitly do "dotnet tool update". But, given it doesn't, it's good to have both here, so that even folks with it already installed get an update.a

danmoseley commented on docs/workflow/ci/triaging-failures.md:
````diff
@@ -0,0 +1,202 @@
+dotnet tool update --global runfo
+```
+
+Then use the `runfo get-helix-payload` command with the job name and workitem name. You can get these from the `Debug` tab of the
````
worth mentioning that it's also in the top line of the console log? At least, the job ID and work item ID are. But apparently runfo requires the work item name, not ID? ID is a guid.

danmoseley reviewed (approved):
lgtm.
